openwisp-config/files/openwisp.agent

```diff
@@ -801,9 +816,9 @@ update_configuration() {
 		env -i ACTION="config-applied" /sbin/hotplug-call openwisp
 		# store the new checksum as last known checksum
 		cp "$CONFIGURATION_CHECKSUM" "$PERSISTENT_CHECKSUM"
-		report_status "applied"
+		retry_with_backoff report_status "applied"
 	else
-		report_status "error"
+		retry_with_backoff report_status "error"
 	fi
```

Comment:
Something which I had not considered before but I think it's worth adding:

```shell
		retry_with_backoff report_status "applied"
	else
		retry_with_backoff report_status "error"
	fi
	# if reporting of the status fails, let it retry in the next cycle
	if [ "$?" -ne "0" ]; then
		rm -f $CONFIGURATION_CHECKSUM $PERSISTENT_CHECKSUM
	fi
```

We should handle the case in which all the attempts to report the status fail so this operation can be repeated in future cycles until it works.